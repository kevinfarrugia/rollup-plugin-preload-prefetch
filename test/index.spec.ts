import { resolve } from "node:path";

import { OutputAsset, RollupOutput } from "rollup";
import { build } from "vite";
import { test, describe, expect } from "vitest";
import PreloadPrefetch, { PreloadPrefetchOptions } from "../src";

async function buildVite(options: PreloadPrefetchOptions[]) {
  const { output } = (await build({
    root: resolve(__dirname, "./fixtures"),
    plugins: [PreloadPrefetch(options)],
  })) as RollupOutput;

  const { source: indexSource } = output.find(
    (item) => item.fileName === "index.html"
  ) as OutputAsset;

  return indexSource.toString();
}

describe("preloadPrefetch", () => {
  test("should add preload to WOFF2 font files", async () => {
    const options: PreloadPrefetchOptions[] = [
      {
        rel: "preload",
        files: [
          {
            entryMatch: /\.woff2$/,
            attributes: {
              type: "font/woff2",
              as: "font",
              crossorigin: "anonymous",
            },
          },
        ],
        injectTo: "head",
      },
    ];

    await buildVite(options).then((result) => {
      expect(result).toMatchSnapshot();
    });
  });

  test("should add prefetch to dynamically imported file", async () => {
    const options: PreloadPrefetchOptions[] = [
      {
        rel: "prefetch",
        files: [
          {
            entryMatch: /form-validation/,
          },
        ],
        injectTo: "head",
      },
    ];

    await buildVite(options).then((result) => {
      expect(result).toMatchSnapshot();
    });
  });

  test("should add preload to WOFF2 font files at start of <head>", async () => {
    const options: PreloadPrefetchOptions[] = [
      {
        rel: "preload",
        files: [
          {
            entryMatch: /\.woff2$/,
            attributes: {
              type: "font/woff2",
              as: "font",
              crossorigin: "anonymous",
            },
          },
        ],
        injectTo: "head-prepend",
      },
    ];

    await buildVite(options).then((result) => {
      expect(result).toMatchSnapshot();
    });
  });
});
