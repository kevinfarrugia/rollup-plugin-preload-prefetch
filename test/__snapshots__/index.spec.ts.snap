// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`preloadPrefetch > should add prefetch to dynamically imported file 1`] = `
"<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Vite App</title>
    <script type="module" crossorigin src="/assets/index-C5-3278v.js"></script>
    <link rel="stylesheet" crossorigin href="/assets/index-BLu0-TaQ.css">
    <link rel="prefetch" href="/assets/form-validation-DZda6X8H.js">
  </head>
  <body>
    <h1>Test App</h1>
  </body>
</html>"
`;

exports[`preloadPrefetch > should add preload to WOFF2 font files 1`] = `
"<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Vite App</title>
    <script type="module" crossorigin src="/assets/index-C5-3278v.js"></script>
    <link rel="stylesheet" crossorigin href="/assets/index-BLu0-TaQ.css">
    <link rel="preload" href="/assets/Roboto-Bold-OBUL28o9.woff2" type="font/woff2" as="font" crossorigin="anonymous">
    <link rel="preload" href="/assets/Roboto-Medium-DRylU_ql.woff2" type="font/woff2" as="font" crossorigin="anonymous">
  </head>
  <body>
    <h1>Test App</h1>
  </body>
</html>"
`;

exports[`preloadPrefetch > should add preload to WOFF2 font files at start of <head> 1`] = `
"<!DOCTYPE html>
<html lang="en">
  <head>
    <link rel="preload" href="/assets/Roboto-Bold-OBUL28o9.woff2" type="font/woff2" as="font" crossorigin="anonymous">
    <link rel="preload" href="/assets/Roboto-Medium-DRylU_ql.woff2" type="font/woff2" as="font" crossorigin="anonymous">

    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Vite App</title>
    <script type="module" crossorigin src="/assets/index-C5-3278v.js"></script>
    <link rel="stylesheet" crossorigin href="/assets/index-BLu0-TaQ.css">
  </head>
  <body>
    <h1>Test App</h1>
  </body>
</html>"
`;
